src/Temporalio/Worker/WorkflowInstance.cs

```diff
@@ -0,0 +1,503 @@
+            {
+                Logger.LogDebug(e, "Workflow raised unexpected failure with run ID {RunID}", Info.RunID);
+                // All exceptions this far fail the task
+                currentActivationException = e;
+            }
+        }
+
+        private void Apply(WorkflowActivationJob job)
+        {
+            switch (job.VariantCase)
+            {
+                case WorkflowActivationJob.VariantOneofCase.CancelWorkflow:
+                    // TODO(cretz): This is only here for now to stop analyzer check error saying we
+                    // never assign to this property. Fix up when building cancel infra.
+                    cancelRequested = true;
+                    break;
+                case WorkflowActivationJob.VariantOneofCase.RemoveFromCache:
+                    // Ignore, handled outside the instance
+                    break;
+                case WorkflowActivationJob.VariantOneofCase.StartWorkflow:
+                    // Don't need parameter, already in instance info
+                    ApplyStartWorkflow();
+                    break;
+                default:
+                    throw new InvalidOperationException($"Unrecognized job: {job.VariantCase}");
+            }
+        }
+
+        private void ApplyStartWorkflow()
+        {
+            QueueNewTask(() => RunTopLevelAsync(async () =>
+            {
+                var resultObj = await Inbound.ExecuteWorkflowAsync(new(
+                    Instance: Instance,
+                    RunMethod: details.Definition.RunMethod,
+                    Parameters: StartParameters,
+                    Headers: details.Start.Headers)).ConfigureAwait(false);
+                var result = details.PayloadConverter.ToPayload(resultObj);
+                AddCommand(new() { CompleteWorkflowExecution = new() { Result = result } });
+            }));
+        }
+
+        /// <summary>
+        /// Workflow inbound implementation.
+        /// </summary>
+        internal class InboundImpl : WorkflowInboundInterceptor
+        {
+            /// <summary>
+            /// Gets the outbound implementation.
+            /// </summary>
+            internal WorkflowOutboundInterceptor? Outbound { get; private set; }
+
+            /// <inheritdoc />
+            public override void Init(WorkflowOutboundInterceptor outbound)
+            {
+                Outbound = outbound;
+            }
+
+            /// <inheritdoc />
+            public override async Task<object?> ExecuteWorkflowAsync(ExecuteWorkflowInput input)
+            {
+                // Have to unwrap and re-throw target invocation exception if present
+                Task resultTask;
+                try
+                {
+                    resultTask = (Task)input.RunMethod.Invoke(input.Instance, input.Parameters)!;
+                }
+                catch (TargetInvocationException e)
+                {
+                    throw e.InnerException!;
```

Comment:
```suggestion
                    ExceptionDispatchInfo.Capture(e.InnerException!).Throw();
```
Properly keep stack trace.

Reply:
Fixed in next PR